Environment shells

ricklamers opened the pull request:
## Description

This feature introduces the concept of "environment shells". It's an interactive shell that can be accessed from the JupyterLab web terminal that is based on one of the project defined environments.

Using this approach users can use the browser based development environment in a more similar style to local development. The terminal gives access to command line tools such as `git` but also data specific CLI programs like `dbt` or `meltano`. In addition, you can test runtime behavior of steps by invoking Python by running `python script.py` in a way that mirrors exactly how the steps will be invoked when part of the pipeline.

Open issue: how to handle step-specific environment variables needed for executing Orchest SDK code (.get_inputs() etc.).
Ideas: make SDK "intelligently" resolve step based on pipeline definition file and current file it's in (problem: file assigned to multiple steps).

Fixes: #938
Fixes: #611

Note, manually tested all base kernels (Julia, JavaScript, Python and R).

🎥 watch me talk through it: https://tella.video/environment-shells-feature-1-fx3w

## Checklist

- [x] I have manually tested my changes and I am happy with the result.
- [x] The documentation reflects the changes.
- [x] The PR branch is set up to merge into `dev` instead of `master`.
- [x] I haven't introduced breaking changes that would disrupt existing jobs, i.e. backwards compatibility is maintained.
- [x] In case I changed code in the `orchest-sdk` I followed its [release checklist](https://github.com/orchest/orchest/blob/master/orchest-sdk/python/RELEASE-CHECKLIST.md).

ricklamers commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -339,11 +463,46 @@ def _get_jupyter_server_deployment_service_manifest(
                             "imagePullPolicy": "IfNotPresent",
                             "volumeMounts": [
                                 volume_mounts_dict["project-dir"],
+                                # TODO: eval (Jacopo): why is the
```
@fruttasecca

yannickperrenet replied:
I believe this has to do with how Docker works. By mounting it directly, the pipeline file can be moved without breaking the mount (because Docker uses the inode). 

So it is still very much applicable. Could you change the `TODO` into a `NOTE` explaining the above?

fruttasecca replied:
Indeed that's the reason

ricklamers replied:
I don't see how the pipeline-file is even used in `jupyter-server`?

ricklamers replied:
Not being used, removing.

yannickperrenet commented on docs/source/fundamentals/environments.md:
````diff
@@ -103,6 +103,22 @@ Installing packages with `conda` is also supported but might take longer (due to
+```
+
+```{warning}
+If you make any changes to your environment, you'll need to restart your environment shell to see the changes. You can restart your environment shell by exiting out of it by typing Ctrl + D or typing 'exit'.
````
```suggestion
If you make any changes to your environment, you'll need to restart your environment shell to see the changes. You can restart your environment shell by exiting out of it by pressing 'Ctrl + D' or typing 'exit'.
```

yannickperrenet commented on docs/source/fundamentals/environments.md:
```diff
@@ -103,6 +103,22 @@ Installing packages with `conda` is also supported but might take longer (due to
 
+## Environment shells
+
+JupyterLab has integrated support for shell sessions using a terminal emulator. To run your scripts and CLI commands while developing your data pipeline you can make use of shells that are based on the environments you have in your project. The environment container image is used such that you have access to all installed programs/dependencies.
```
We choose to capitalize all terms in Orchest, so `environment` should be `Environment` ([our implementation detail docs](https://docs.orchest.io/en/stable/development/implementation_best_practices.html#writing-docs)). 

```suggestion
JupyterLab has integrated support for shell sessions using a terminal emulator. To run your scripts and CLI commands while developing your data Pipeline you can make use of shells that are based on the Environments you have in your Project. The Environment container image is used such that you have access to all installed programs/dependencies.
```

yannickperrenet commented on docs/source/fundamentals/environments.md:
```diff
@@ -103,6 +103,22 @@ Installing packages with `conda` is also supported but might take longer (due to
+
+JupyterLab has integrated support for shell sessions using a terminal emulator. To run your scripts and CLI commands while developing your data pipeline you can make use of shells that are based on the environments you have in your project. The environment container image is used such that you have access to all installed programs/dependencies.
+
+All you need to do is open JupyterLab in a session and click on Terminal:
```
```suggestion
All you need to do is open JupyterLab and click on Terminal:
```

Don't think `in a session` adds anything here (especially given that we start the Session automatically).

yannickperrenet commented on lib/python/orchest-internals/_orchest/internals/utils.py:
```diff
@@ -128,6 +128,10 @@ def get_init_container_manifest(
     init_container = {
         "name": "image-puller",
         "image": image_puller_image,
+        "securityContext": {
+            "privileged": True,
+            "runAsUser": 0,
+        },
```
Why was this added in the context of Environment shells?

ricklamers replied:
Ah good catch, this was a left over of me experimenting with running the user containers (steps, shells, kernel) as UID 1000. The init container needed securityContext runAsUser 0/priviliged true to be able to still use the socket for its work. Since we're not changing to UID 1000 because of root owned disk state in production I can drop this as the parent container is still running as UID 0.

ricklamers replied:
On second thought, I think it might be worth it leaving this in. We'd not be implicitly depending on the parent container of any container that includes the init container to be running as privileged/root. Wdyt?

yannickperrenet commented on orchest-sdk/python/orchest/utils.py:
```diff
@@ -8,6 +8,22 @@
 from orchest.pipeline import Pipeline
 
 
+def pretty_print_env_var_suggestion(pipeline: Pipeline):
```
```suggestion
def pretty_print_env_var_suggestion(pipeline: Pipeline) -> None:
```

Common practice to explicitly return `None` if applicable.

yannickperrenet commented on orchest-sdk/python/orchest/utils.py:
```diff
@@ -8,6 +8,22 @@
+        + "environment variable. Choose one of the following pipeline steps:"
+    )
+    for step in pipeline.steps:
+        print(
+            "ORCHEST_STEP_UUID="
+            + step.properties.get("uuid")
+            + " (Step %s %s)"
+            % (step.properties.get("title"), step.properties.get("file_path"))
+        )
```
Nitpick (feel free to choose whatever): in my opinion readability of fstrings is better here.
```suggestion
        print(
            f'ORCHEST_STEP_UUID={step.properties.get("uuid")} '
            f'(Step {step.properties.get("title")}, {step.properties.get("file_path")})'
        )
```

(Wanted to add it to show how fstring usage would look.)

ricklamers replied:
Haha still not convinced in this case :D

yannickperrenet commented on orchest-sdk/python/orchest/utils.py:
```diff
@@ -8,6 +8,22 @@
+        print(
+            "ORCHEST_STEP_UUID="
+            + step.properties.get("uuid")
+            + " (Step %s %s)"
```
```suggestion
            + " (Step: '%s' at '%s')"
```

yannickperrenet commented on orchest-sdk/python/orchest/utils.py:
```diff
@@ -30,7 +46,11 @@ def get_step_uuid(pipeline: Pipeline) -> str:
+        # an environment shell. Print environment variable setter hint.
+        pretty_print_env_var_suggestion(pipeline)
+
+        raise StepUUIDResolveError("Could not resolve the pipeline step UUID.")
```
Given that the error is called `StepUUIDResolveError` I think it is already clear that the "pipeline step UUID couldn't be resolved". Therefore, I think this gives us the opportunity to use the message we can pass with the error to better explain what happened, e.g.:

```suggestion
        raise StepUUIDResolveError(
            'Environment variable "STEP_UUID" has to be explicitly passed when'
            ' executing Pipeline Steps in Environment shells.'
        )
```

yannickperrenet commented on services/base-images/base-kernel-javascript/Dockerfile:
```diff
@@ -27,9 +27,12 @@ COPY ./runnable-shared/runner/requirements* /orchest/services/base-images/runnab
 COPY ./lib/python /orchest/lib/python
 COPY ./orchest-sdk /orchest/orchest-sdk
 RUN chown $NB_USER -R /orchest/orchest-sdk \
-    && chown $NB_USER -R /orchest/lib
+    && chown $NB_USER -R /orchest/lib && \
+    echo -e "jovyan\njovyan" | passwd jovyan
 
+# Exclude hostname from default prompt
 USER $NB_USER
+RUN echo "export PS1='\[\033[01;32m\]\u\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '" >> ~/.bashrc
```
To reduce the number of layers in the Dockerfile (and use consistent way of breaking `RUN` commands over multiple lines):
```suggestion
RUN chown $NB_USER -R /orchest/orchest-sdk \
    && chown $NB_USER -R /orchest/lib \
    && echo -e "jovyan\njovyan" | passwd jovyan \ 
    # Exclude hostname from default prompt
    && echo "export PS1='\[\033[01;32m\]\u\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '" >> /home/$NB_USER/.bashrc

USER $NB_USER
```

yannickperrenet commented on services/base-images/runnable-shared/bootscript.sh:
```diff
@@ -10,6 +10,14 @@ umask 002
+    # Mark safe for git
+    git config --global --add safe.directory /project-dir
+    
+    env > /etc/environment
+    sudo service ssh start
+    sudo service ssh stop
+    sudo /usr/sbin/sshd -D
```
Why is the `ssh` service started and then immediately stopped?

fruttasecca replied:
If this is intended it might be worth it to add a comment in there

yannickperrenet commented on services/jupyter-server/Dockerfile:
```diff
@@ -41,6 +41,12 @@ RUN bash /patch-pip.sh
+COPY shellspawner/shellspawner /usr/bin/shellspawner
+
+# Add shellspawner to terminal profile
+# TODO: check if this doesn't get sourced for the main start.sh
```
I guess this `TODO` is no longer applicable?

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+
+def print_jupyter_warning():
+    print(
+        "\nUsing the JupyterLab terminal directly is not recommended "
+        + "read more about it in the docs\n"
+        + "➜ https://docs.orchest.io/en/stable/fundamentals/environments.html"
```
Nitpick: Using `+` is not needed actually.
```suggestion
        "\nUsing the JupyterLab terminal directly is not recommended"
        " read more about it in the docs\n"
        "➜ https://docs.orchest.io/en/stable/fundamentals/environments.html"
```

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+
+
+def environment_shell_uuid_to_environment_uuid(environment_shell_uuid: str) -> str:
+    random_uuid_suffix_len = 6
```
Might be better to define a global constant given that the suffix is used in multiple functions

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+def environment_shell_uuid_to_name(
+    environment_shell_uuid: str, environments: List[Dict[str, Any]]
+) -> str:
+    random_uuid_suffix_len = 6
+
+    # Cut of last random UUID
+    environment_uuid = environment_shell_uuid[: -(random_uuid_suffix_len + 1)]
```
Why not use the dedicated function `environment_shell_uuid_to_environment_uuid` to do this?

yannickperrenet replied:
Could even make it into `def split_environment_shell_uuid(...) -> Tuple[str, str]` which returns the `environment_uuid` and `shell_name`.

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+            print(f"{prefix} [0{choice_end}]: ", end="", flush=True)
+            user_input = input()
+            index = int(user_input)
+            assert index >= 0 and index <= choice_count - 1
```
I think it would be better to check with an if-statement and raise a `ValueError` given that it is user input we are testing. In addition assert statements (see [python docs](https://docs.python.org/3/reference/simple_stmts.html#assert)) are used as a debugging tool and can actually be disabled with a flag, e.g. `python -O script.py`. Given that our implementation is actually (in prod) dependent on the assert statement running, I think it is better to be explicit and make sure the program fails if we want it to.

```suggestion
            if not (index >= 0 and index <= choice_count - 1):
                raise ValueError("Invalid user input.")
```

fruttasecca replied:
>I think it would be better to check with an if-statement and raise a ValueError given that it is user input we are testing

Agree

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+            if choice_count == 1:
+                return 0
+
+            print("Invalid input.")
```
```suggestion
            print(f"Invalid input. Enter integer between 0 and {choice_count - 1}.")
```

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+            continue
+
+
+def get_environments() -> List[Dict[str, Any]]:
+    req = request.Request(
+        ORCHEST_WEBSERVER_ADDRESS
+        + "/store/environments/"
+        + os.environ["ORCHEST_PROJECT_UUID"]
+    )
+    req.add_header("Content-Type", "application/json")
```
Given that Environment shells actually require Environments to have been built, wouldn't it be better to query the `orchest-api` instead? Otherwise we can run into the issue where a user requests an Environment shell but the Environment not being available yet.

(On top of my head, without actually testing it) a reproducible test case:
1. Start a session
2. Open jupyterlab
3. Open Orchest in another tab
4. Create a new Environment (but don't build it)
5. `orchest-websever` will now return an Environment that doesn't actually have an Environment image yet.

fruttasecca replied:
Agree with this, it could also be the case that an old environment image doesn't support shells given that it doesn't have the ssh server. The current surface of the orchest-api allows to know all this information. Also the orchest-webserver is not the right entity to ask for such information regardless, imo, this definitively feels like responsibility of the orchest-api.

ricklamers replied:
So `orchest-api` doesn't have the environment name information which is needed to show the list to the user.

The assumption we use (because of the gate) that no environment is unavailable if the user is able to start a session (that runs JupyterLab). This guarantee is not currently maintained in a multi-tab context anyway (irrespective of the environment shells feature).

As long as environment names aren't managed at the `orchest-api` level I think using the `orchest-webserver` for fetching the canonical list of environments I think makes sense.

I'm also not a fan of having "another gate" at the level of environment shells. I think that should be handled at the interactive session level which is a required thing to be running when you want to start an environment shell.

yannickperrenet replied:
> The assumption we use (because of the gate) that no environment is unavailable if the user is able to start a session (that runs JupyterLab).

Could you add that as a comment (@ricklamers)? I think it is a great explanation as to why the webserver is queried, not the orchest-api.

yannickperrenet replied:
https://github.com/orchest/orchest/pull/1248/commits/e5476e0ddea924fce8545291707f15f8e5445701

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+        wait_index += 1
+
+    print("")
+    if waited >= timeout:
+        print("Warning! Loader timed out after %d seconds." % timeout)
```
Shouldn't this be an error instead? I guess if this case is hit then the Environment shell won't connect?

If that is the case then this print statement is a little misleading for users.

ricklamers replied:
I think throwing an error is the responsibility of the check_ready_callable not of the loader.

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+    return json_resp
+
+
+def spawn_environment_shell(environment: Dict[str, Any]) -> Dict[str, Any]:
+    t = BackgroundRequest(lambda x: _spawn_environment_shell(x), [environment])
+    t.start()
+    show_loader(lambda x: x.request_ready, [t], "Starting shell")
+    return t.request_result
```
Ghehe nice! This pattern is actually very similar to using [futures](https://docs.python.org/3/library/asyncio-future.html) in Python: `await`ing the future and displaying a spinner in the meantime.

ricklamers replied:
:D

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+def connect(environment_shell: str):
+    print("Connecting...", flush=True)
+
+    user = "jovyan"  # TODO: support more usernames
```
Wouldn't it be better to use the `$NB_USER` environment variable?

ricklamers replied:
I don't want to tightly couple the `jupyter-server` and the base images. $NB_USER doesn't currently exist in the context of `jupyter-server` to I think not introducing that is the best option.

The fact that this username is not generic is more of a MVP scope choice. Similar to how we currently don't have extensive docs/support for custom base images (although with enough work a user could technically bring their own custom images by mimicking what we're doing in our base images to the extent necessary - runner etc.).

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+    init()
+
+
+def choose_environment(environments: List[Dict[str, Any]]) -> Dict[str, Any]:
```
Strictly speaking the return type is actually: `Optional[Dict[str, Any]]`. Given that if `environment_choice` is `None`, then `None` is returned (which is possible).

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+    if action is None:
+        quit()
+
+    if action == 0:
```
This (using `0` hardcoded) closely couples this function with `choose_action`. Wouldn't it be possible to change the action into an `Enum`?

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+    if len(sys.argv) > 1 and sys.argv[1] == "--help":
+        print("This terminal supports connecting with Orchest environment shells.")
+        print("To restart, terminate, start or connect to a shell simply exit out of ")
+        print("your current shell by typing 'exit' or typing Ctrl + D.")
```
```suggestion
        print("your current shell by typing 'exit' or pressing 'Ctrl + D'.")
```

yannickperrenet replied:
@ricklamers Don't see this committed yet? Also fine if you don't disagree with the change of course.

ricklamers replied:
Typing is the generic term for entering keys to a system. Pressing implies a specific entry mechanism (one might use the mouse to 'type' Ctrl + D on some system). The quotes I think is more sensible for text because it's a literal string. Whereas Ctrl and D just refer to things, so I didn't think quoting them was necessary.

I'm :100: overthinking this.

yannickperrenet commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+        print("This terminal supports connecting with Orchest environment shells.")
+        print("To restart, terminate, start or connect to a shell simply exit out of ")
+        print("your current shell by typing 'exit' or typing Ctrl + D.")
+        print("")
```
Total nitpick:
```suggestion
        print()
```

yannickperrenet commented on services/orchest-api/app/app/schema.py:
```diff
@@ -239,6 +239,43 @@ def _session_base_url(s) -> str:
+    {
+        "uuid": fields.String(required=True, description="UUID of environment shell"),
+        "session_uuid": fields.String(required=True, description="UUID of session"),
+        "host": fields.String(required=True, description="host of k8s service"),
```
I am not sure I understand `"host of k8s service"`. What do you mean here?

ricklamers replied:
I've updated it to `hostname of environment shell in k8s cluster`. Does that make more sense?

I want to avoid baking "k8s concepts" into the domain model of the environment shell. `hostname` feels more generic for future use.

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -339,11 +463,46 @@ def _get_jupyter_server_deployment_service_manifest(
+                                    "name": "ORCHEST_PIPELINE_UUID",
+                                    "value": session_config["pipeline_uuid"],
+                                },
+                                # Note, this is not the
+                                # ORCHEST_PIPELINE_PATH
+                                # that's used by the SDK that points
+                                # to the absolute path of the mounted
+                                # pipeline file.
+                                # ORCHEST_SESSION_PIPELINE_PATH is used
+                                # to let the shellspawner pass the
+                                # relative pipeline path to the
+                                # environment-shells endpoint.
+                                {
+                                    "name": "ORCHEST_SESSION_PIPELINE_PATH",
+                                    "value": session_config["pipeline_path"],
+                                },
```
How about changing the name to `ORCHEST_SHELLSPAWNER_PIPELINE_PATH`? If you have to explain the name, then it is probably not a good one :P

ricklamers replied:
I think the environment variable should be named to reflect the content not the target program for which the environment variable is intended. It's the pipeline path for the current session. Since we have another similar name that is the actual path (on the system, the one used by the SDK to read the pipeline file at that location) I thought it would be good to describe the difference to avoid confusion.

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -23,7 +24,7 @@
     userdir_pvc: str,
     project_dir: str,
-    pipeline_path: str,
+    pipeline_path: str = None,
```
If we set it to `None` by default then the type needs to be `Optional[str]`.

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +279,139 @@ def _get_session_sidecar_deployment_manifest(
     }
 
 
+def get_orchest_sdk_vars(
```
Imo better to make this a private function, i.e. `_get_orchest_sdk_vars`

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +279,139 @@ def _get_session_sidecar_deployment_manifest(
+    project_dir: str,
+    environment_image: str,
+) -> Tuple[Dict[str, Any], Dict[str, Any], Dict[str, Any]]:
+    """
+    This manifest generation is in core/sessions
+    since the environment shell is part of the
+    session, but has a detached lifecycle in the sense
+    that environment shells can be started/stopped
+    independently from the session start/stop.
+
+    When a session is stopped all associated environment
+    shells should always be stopped.
+    """
```
Nitpick: Docstring doesn't seem to be wrapped at `72` chars.

fruttasecca replied:
Is this comment actually needed? It seems to suggest that some action needs to be taken to clean up things while the session shutdown logic will already take care of it AFAIK

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +279,139 @@ def _get_session_sidecar_deployment_manifest(
+    """
+
+    metadata = {
+        "name": f"environment-shell-{environment_uuid}-{(str(uuid4()))[:6]}",
```
Can't we make this `6` a config option in the internal lib? I have seen this come by a few times now and I am scared we could forget updating it in one of its places.

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +279,139 @@ def _get_session_sidecar_deployment_manifest(
+
+    metadata = {
+        "name": f"environment-shell-{environment_uuid}-{(str(uuid4()))[:6]}",
+        "labels": {
+            "app": "environment-shell",
+            "project_uuid": project_uuid,
+            "session_uuid": session_uuid,
+        },
```
How about also adding `shell_uuid` here? The value can be the generated uuid4 from above (this way we also store the full uuid instead of just the trimmed (to 6 chars) version).

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +279,139 @@ def _get_session_sidecar_deployment_manifest(
+        "metadata": metadata,
+        "spec": {
+            "type": "ClusterIP",
+            "selector": metadata["labels"],
```
Continuation on https://github.com/orchest/orchest/pull/1248/files#r968048749, I think (without specifying the shell_uuid in the labels) that this service would point to all Environment shells (given that all shells within the same session have the same uuid). Meaning that if a user starts many shells, then kubernetes will choose what shell to route traffic to.

ricklamers replied:
> given that all shells within the same session have the same uuid

They don't?

yannickperrenet replied:
I guess I was not clear the first time, what I meant to say is that the `metadata["labels"]` (with the current implementation) is not unique for different environment shells. This will lead to different environment shell k8s services will point to the same Pods as backends, meaning that the different services point to the same pods. This in turn means that as a user you might connect to a different shell than you are expected.

![image](https://user-images.githubusercontent.com/26223174/189645171-97995d17-9437-4333-8eac-ad549129d3f4.png)

So the `metadata["labels"]` need to include the full `shell_uuid`.

ricklamers replied:
This should now be fixed given we include a `shell_uuid` label

yannickperrenet commented on services/orchest-webserver/app/app/views/orchest_api.py:
```diff
@@ -351,6 +353,32 @@ def catch_api_proxy_sessions_post():
+                "project_dir": project_dir,
+            }
+
+            # Don't wait for the response
+            threading.Thread(
+                target=requests.post,
+                args=(
+                    "http://"
+                    + app.config["ORCHEST_API_ADDRESS"]
+                    + "/api/environment-shells/",
+                ),
+                kwargs={"json": json_data},
+            ).start()
```
This should probably use the `current_app.config["SCHEDULER"]` instead (to stay consistent with the rest of the codebase). Example: https://github.com/orchest/orchest/blob/377f596e915db5c03aa474f60dedb3fe5b5bf0c2/services/orchest-webserver/app/app/core/projects.py#L196-L201

yannickperrenet commented on services/orchest-webserver/app/app/views/orchest_api.py:
```diff
@@ -351,6 +353,32 @@ def catch_api_proxy_sessions_post():
+                "pipeline_uuid": pipeline_uuid,
+                "pipeline_path": pipeline_path,
+                "project_uuid": project_uuid,
+                "environment_uuid": project_envs[0].uuid,
```
Might be good explaining (in a comment) why environment `[0]` is used.

yannickperrenet commented on services/orchest-webserver/app/app/views/orchest_api.py:
```diff
@@ -351,6 +353,32 @@ def catch_api_proxy_sessions_post():
+        # Side effect of a session post is to create an environment
+        # shell this is a "client decision" hence it's not handled
+        # automatically by the orchest-api.
+        project_envs = get_environments(project_uuid)
```
I think this has the same problem as explained in https://github.com/orchest/orchest/pull/1248#discussion_r967153695.

yannickperrenet commented on services/orchest-api/app/app/core/environment_shells.py:
```diff
@@ -0,0 +1,151 @@
+    """
+
+    environment_shell_service_k8s_deployment_manifests = []
+    environment_shell_service_manifest = []
+
+    (depl, serv,) = _manifests._get_environment_shell_deployment_service_manifest(
+        session_uuid,
+        environment_uuid,
+        project_uuid,
+        pipeline_uuid,
+        pipeline_path,
+        userdir_pvc,
+        project_dir,
+        environment_image,
+    )
+    environment_shell_service_manifest = serv
```
This is a redefinition of `environment_shell_service_manifest`, are you sure you didn't want to append the `serv` to the list?

yannickperrenet commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+            + f":{environment_image.tag}"
+        )
+
+        return launch_environment_shell(
+            session_uuid,
+            environment_shell_config["environment_uuid"],
+            environment_shell_config["project_uuid"],
+            environment_shell_config["pipeline_uuid"],
+            environment_shell_config["pipeline_path"],
+            environment_shell_config["userdir_pvc"],
+            environment_shell_config["project_dir"],
+            environment_image_string,
+        )
```
In my opinion, this should be placed in the `_collateral` in order to split DB calls from infra changes (as is the idea behind the `TwoPhaseFunction`)

fruttasecca replied:
Agree

fruttasecca commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+#!/usr/bin/env python3
+
+# Zero dependency shell spawner program
+# Connects JupyterLab terminal to an SSH daemon
+# running on an environment shell container.
```
Might be preferable to write this as a module docstring

fruttasecca commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+    environment_shells: List[Dict[str, Any]],
+    environments: List[Dict[str, Any]],
+    prompt_suffix: str,
+) -> Dict[str, Any]:
```
This function can actually return `None`

fruttasecca commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+        session_uuid/environment_shell_uuid."""
+        try:
+            stop_environment_shell(environment_shell_uuid)
+            return {}, 201
```
```suggestion
            return {}, 200
```

fruttasecca commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+class EnvironmentShellList(Resource):
+    @api.doc("launch_environment_shell")
+    @api.expect(schema.environment_shell_config)
+    @api.marshal_with(schema.environment_shell, code=200)
```
```suggestion
    @api.marshal_with(schema.environment_shell, code=201)
```

fruttasecca commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+            pipeline_uuid=environment_shell_config["pipeline_uuid"],
+        ).one_or_none()
+        if isess is None:
+            return {"message": "Session doesn't exist."}, 409
```
```suggestion
            return {"message": "Session doesn't exist."}, 404
```

fruttasecca commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+        if isess is None:
+            return {"message": "Session doesn't exist."}, 409
+
+        environment_shell = None
```
Is this line needed?

fruttasecca commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+        return environment_shell
+
+
+class CreateEnvironmentShell(TwoPhaseFunction):
+    def _transaction(self, environment_shell_config) -> Dict[str, Any]:
```
In the transaction `environments.lock_environment_images_for_interactive_session` should be called to make sure the `orchest-api` knowns that the image is in use by the session

ricklamers replied:
Isn't that handled by the interactive session that the environment shells always depend on?

fruttasecca replied:
Not entirely because while the session is active you could build another image then start an env shell. See `/kernels/lock-image/` in `services/orchest-api/app/app/apis/namespace_sessions.py`. This is needed to allow users to update an environment image and get an up to date kernel/shell without restarting the session

fruttasecca commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+from typing import Any, Dict
+
+from flask import request
+from flask.globals import current_app
```
Imo a module-level docstring explaining the module scope, purpose, usage etc. will have a good ROI

fruttasecca commented on services/orchest-api/app/app/core/environment_shells.py:
```diff
@@ -0,0 +1,151 @@
+from _orchest.internals import config as _config
+from app import utils
+from app.connections import k8s_apps_api, k8s_core_api
+from app.core.sessions import _manifests
```
Imo the fact that we need to access a "private" module of sessions is a sign that the taxonomy needs to be slightly altered. What I mean with this is that, as it is, environment shells are strictly encapsulated in interactive sessions, but not all code reflects that (like in this case). 

Imo we should reflect if environment shells strictly being part of an interactive session is desirable, meaning that we might choose to have environment-based shells without a backing interactive session in the future, and, if that's the case, the cost to implement that will become higher if the current implementation strongly assumes that env shells are part of an interactive session.

If env shells are part of interactive sessions then this module should probably be moved into the `app.core.sessions` directory

ricklamers replied:
The way I see it, currently we want the environment shell to be strictly dependent on interactive sessions but we may move out of that in the future. Hence I prefer "weakly" coupling them like I do now to make that migration in the future less disruptive (no need to "migrate" to new endpoints without the /sessions/  prefix for example).

The context in which I see us becoming independent of interactive sessions in the future is with remote IDE support.

fruttasecca commented on services/orchest-api/app/app/core/environment_shells.py:
```diff
@@ -0,0 +1,151 @@
+    """Starts environment shell
+
+    Args:
+        session_uuid: UUID to identify the session k8s namespace with,
+            which is where all related resources will be deployed.
+        environment_uuid: envirohnment UUID to embed UUID in the service
+            name which is used for pretty-printing and restarts.
+        project_uuid: UUID of the project.
+        pipeline_uuid: UUID of the pipeline.
+        pipeline_path: Relative path (from project directory root) to
+            the pipeline file e.g. 'abc/pipeline.orchest'.
+        userdir_pvc: Name of the k8s PVC e.g. 'userdir-pvc'.
+        project_dir: Name of the project directory e.g. 'my-project'
+            note this is always a single path component.
+        environment_image: The full image specification that can be
+            given directly as the image string to the container runtime.
```
The fact that an environment shell depends on a pipeline is a bit strange, I know that this information is needed somewhere but I feel like that the amount of information needed to start, essentially, a pod could be reduced. From a far-off (and perhaps naive) point of view one could say that only the environment, project and session uuids are really needed

ricklamers replied:
This is an intended dependency, the environment shell is dependent on a specific pipeline in order to make the `orchest-sdk` function properly.

If we want to make environment shells independent of the pipeline path we'd first need to change how the `orchest-sdk` works which is out of scope imo.

fruttasecca commented on services/orchest-api/app/app/core/environment_shells.py:
```diff
@@ -0,0 +1,151 @@
+    Will be cleaned up when the session is stopped.
+    """
+
+    environment_shell_service_k8s_deployment_manifests = []
+    environment_shell_service_manifest = []
```
Are lists actually needed here?

fruttasecca commented on services/orchest-api/app/app/core/environment_shells.py:
```diff
@@ -0,0 +1,151 @@
+
+    ns = _config.ORCHEST_NAMESPACE
+
+    logger.info("Creating environment shell services deployments.")
```
Related to https://github.com/orchest/orchest/pull/1248/files#r968124248, this function only creates 1 deployment and service right?

fruttasecca commented on services/orchest-api/app/app/core/environment_shells.py:
```diff
@@ -0,0 +1,151 @@
+      pods
+      service_accounts
+      role_bindings
+      roles
+
+    Will be cleaned up when the session is stopped.
```
Imo the docstring should specify what's being returned

ricklamers replied:
Removed the return as part of the refactor that moves the launch_ call to collateral of the two phase function.

fruttasecca commented on services/orchest-api/app/app/core/environment_shells.py:
```diff
@@ -0,0 +1,151 @@
+    }
+
+
+def get_environment_shells(session_uuid):
```
Imo typing should be added for consistency with the rest of the module

fruttasecca commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -246,27 +250,14 @@ def _get_session_sidecar_deployment_manifest(
-                                    "name": "ORCHEST_SESSION_TYPE",
-                                    "value": session_type,
-                                },
+                            "env": get_orchest_sdk_vars(
+                                project_uuid,
+                                pipeline_uuid,
+                                _config.PIPELINE_FILE,
+                                session_uuid,
+                                session_type,
+                            )
```
Given that we are talking about the session sidecar, which doesn't use the orchest SDK, I'm not sure the name of the function fits

ricklamers replied:
I wanted to make it more explicit which environment variables are part of the SDK to not repeat that everywhere these are injected.

Maybe it's the wrong layer of abstraction. Do you have an idea that isn't repeating the list of environment variables needed for the SDK everywhere?

fruttasecca replied:
Not sure tbh, let's leave it as is

fruttasecca commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +279,139 @@ def _get_session_sidecar_deployment_manifest(
 
 
+def get_orchest_sdk_vars(
+    project_uuid, pipeline_uuid, pipeline_file, session_uuid, session_type
+):
```
Note that the `services/orchest-api/app/app/types.py` module has a `SessionType` enum

fruttasecca commented:
A bit meta but the PR description could/should have contained a small description of the architectural decisions behind the changes, e.g. "an env shell lifecycle can be different than a session", etc..

I think this particular feature deserves an entry in the "Implementation details" docs

fruttasecca commented:
A thing to keep in mind is that the current definition of orchest being idle in `services/orchest-api/app/app/apis/namespace_info.py` `is_orchest_idle` includes busy kernels from interactive sessions but doesn't include "busy" environment shells (not even sure we can detect that)

fruttasecca commented:
I think the cleanup logic in `services/orchest-api/app/app/__init__.py`, `cleanup()` needs to be incremented to include environment shells that are detached from interactive sessions

fruttasecca commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +279,139 @@ def _get_session_sidecar_deployment_manifest(
+    environment_image: str,
+) -> Tuple[Dict[str, Any], Dict[str, Any], Dict[str, Any]]:
+    """
+    This manifest generation is in core/sessions
+    since the environment shell is part of the
+    session, but has a detached lifecycle in the sense
+    that environment shells can be started/stopped
+    independently from the session start/stop.
```
I was slightly confused by this comment, I initially interpreted it in a way that suggested that env shells could be started independently of interactive sessions, while it's actually that the lifetime of env shells is within the bounds of the interactive session, env shells are started independently within these bounds but are always stopped when the session is stopped.

ricklamers replied:
Good point, I'll make the comment more reflective of that fact.

ricklamers commented on services/jupyter-server/shellspawner/shellspawner:
```diff
@@ -0,0 +1,371 @@
+            self.request_result = self.request_callable(*self.callable_args)
+            self.request_ready = True
+        except (urllib.error.HTTPError, urllib.error.URLError) as e:
+            print(e)
```
Makes sense, but in that case I think the entire try/except isn't necessary. Removed it

fruttasecca commented on services/orchest-api/app/app/apis/namespace_environment_shells.py:
```diff
@@ -0,0 +1,115 @@
+api = schema.register_schema(api)
+
+
+@api.route("/<string:session_uuid>")
```
Perhaps the API endpoints could be improved/made a bit more restful. Currently, we have the following:

- GET `/environment-shells/<uuid>` -> gets all shells of a session
- POST `/environment-shells` -> creates a new shell in a certain *session* which is implicitly inferred by the proj/ppl uuid
- DELETE `/environment-shells/<uuid>` -> deletes an env shell by uuid

So a uuid as the first value after `/environment-shells` is seen as a session uuid in GETs and as a shell uuid in DELETEs. This comment is related to https://github.com/orchest/orchest/pull/1248#discussion_r968114137, the taxonomy and the resulting API endpoints paths should be slightly adjusted imo, to better reflect the structure/architecture

ricklamers commented:
> A thing to keep in mind is that the current definition of orchest being idle in `services/orchest-api/app/app/apis/namespace_info.py` `is_orchest_idle` includes busy kernels from interactive sessions but doesn't include "busy" environment shells (not even sure we can detect that)

Good point but out of scope imo. At least for now.

ricklamers commented:
> A bit meta but the PR description could/should have contained a small description of the architectural decisions behind the changes, e.g. "an env shell lifecycle can be different than a session", etc..
> 
> I think this particular feature deserves an entry in the "Implementation details" docs

Anything in particular you'd like to see documented there?

yannickperrenet commented:
Okay, now that we have the code review out of the way, I have some high level comments (that I ran into when testing the PR).

----

![image](https://user-images.githubusercontent.com/26223174/189703569-dfcc1091-594d-4759-ac61-8e24fb8811d1.png)

- [ ] I think the above message can be improved. As a new user I have no idea what Environment shells are and when seeing this message I will just `exit` instead of realizing I am already in an Environment shell. Because it reads "to ... connect to a shell simply exit out of your current shell". How about changing it to "connect to **another** shell" (without the bold of course).  

----

![image](https://user-images.githubusercontent.com/26223174/189704551-d4515ae8-a388-4e09-93c5-e147a18b49d5.png)

- [ ] On `Ctrl + C` (since we catch the error) there should be an empty line print to have the `Choose option` on a new line.
- [ ] How about catching the `ValueError` and instead just exit the choice-menu silently?
- [ ] I guess the only way for me to now get back into an Environment shell is to start another JupyterLab Terminal? This would be fine with me btw.

----

![image](https://user-images.githubusercontent.com/26223174/189705416-3bca03cc-41b5-4cd0-b655-de847051aa1d.png)

- [ ] Can we make it even clearer to users that they are in an Environment shell right now? How about doing something similar to activated virtual environment indications (what many people configure their shells to do, e.g. `oh-my-zsh`): 
   `jovyan:/project-dir (Environment shell: Python-3 223hdf) $ ll`

----

![image](https://user-images.githubusercontent.com/26223174/189706077-20f44052-53ab-4a72-bf35-24cb0abf541c.png)

- [ ] Nitpick: Do we want another option to `Exit out of menu`? Otherwise only `Ctrl + C` gets a user to the JupyterLab terminal. Is it on purpose that it is so hard? I think another option in the menu works (given that the help message `exit` or `Ctrl + D` only get "Invalid input").

yannickperrenet commented on services/orchest-api/app/app/core/sessions/_manifests.py:
```diff
@@ -288,6 +278,146 @@ def _get_session_sidecar_deployment_manifest(
+    all associated environment shells.
+    """
+
+    metadata = {
+        "name": service_name,
+        "labels": {
+            "app": "environment-shell",
+            "project_uuid": project_uuid,
+            "session_uuid": session_uuid,
+            "shell_uuid": session_uuid,
+        },
+    }
```
:exclamation: `shell_uuid` should **not** map to `session_uuid` here. It has to point to a uuid unique to the shell, otherwise we run into the issue of 1 service pointing to multiple pods (which is still what happens currently).

ricklamers replied:
Woop that's a misatek

ricklamers commented:
>  I guess the only way for me to now get back into an Environment shell is to start another JupyterLab Terminal? This would be fine with me btw.

I've added this to a print message. Technically they could also invoke `shellspawner` but I don't want to expose that detail to the user. See my remark below about not wanting to guide toward JupyterLab terminal.

> Nitpick: Do we want another option to Exit out of menu? Otherwise only Ctrl + C gets a user to the JupyterLab terminal. Is it on purpose that it is so hard? I think another option in the menu works (given that the help message exit or Ctrl + D only get "Invalid input").

I actually left that out on purpose. I want it to feel like "this is not what I'm supposed to do". You can get to the "underlying" JupyterLab shell, but everything is "broken" there because you have none of the environment installed dependencies available. I don't want to block it altogether so leaving it in behind a `Ctrl + C` keyboard interrupt seemed like the best trade-off option.

> Nitpick: How about adding an executable called help that users can invoke to re-print the initial greeting message? Whilst testing I something had to think "Hmmm what did I have to do again?". I feel like I am in the habit of always trying something like help in these types of environments, so I though it could be a good addition here.

That message is part of the `shellspawner` program which does not exist in the base images. A separation I'm keen on keeping (we would have to "ship" shellspawner to the base images just for the help message). In addition, the message is on the screen for every terminal you open in JupyterLab. Furthermore, we'd be overriding the built-in `help` of bash. Which some people may like referring to. So I'm not a fan of including our own `help` override.

Great feedback! What I ended up implementing is checkboxed. And I fixed the label bug.

ricklamers commented:
>I actually left that out on purpose. I want it to feel like "this is not what I'm supposed to do". You can get to the "underlying" JupyterLab shell, but everything is "broken" there because you have none of the environment installed dependencies available. I don't want to block it altogether so leaving it in behind a Ctrl + C keyboard interrupt seemed like the best trade-off option.

Since using `git` from a terminal actually works best from the built-in JupyterLab terminal until we implement `git` ssh key management I think we should leave the option in. This way people who've configured `git` through JupyterLab config (https://docs.orchest.io/en/stable/fundamentals/projects.html#using-git-inside-orchest-projects) can still make use of it through the terminal (the GUI will always work).

yannickperrenet commented:
![image](https://user-images.githubusercontent.com/26223174/189910026-477d7dc0-53de-4a54-8960-b113f3f1ccaf.png)

I think the current indication of the Environment shell is confusing (especially as a new user who doesn't know what Environment shells). How about: `(Environment: Python-2-typer)`?

We can just explain them they are running inside an Environment, doesn't even need to be `Environment shell` (who cares what we name it). Yes, this has the issue of showing the same indication for two different shells for same Environment, but I think that is just fine.

yannickperrenet commented:
![image](https://user-images.githubusercontent.com/26223174/189911015-bb5a4c12-70e9-424b-80a9-f5bffb1339bb.png)

I really like this message when exiting to the JupyterLab terminal! I think it should also be shown when option `[4] Exit to JupyterLab built-in shell` is chosen (currently it just exits and doesn't show this message).

yannickperrenet commented:
![image](https://user-images.githubusercontent.com/26223174/189911595-c0b377b1-215b-40e4-b19b-d1086248cd16.png)

As a user it can be confusing that options can no longer be chosen, because we never indicated that the options are only inside Environment shells. Note that the `jovyan` prompt was never shown.

To reproduce:
1. Start session
2. Open JupyterLab terminal
3. Get auto connected to shell
4. `exit` 
5. Start a new shell
6. Wait until you are connected
7. Close the terminal
8. Start a new terminal
9. Note that the `Connecting...` etc. is never shown

ricklamers commented:
> As a user it can be confusing that options can no longer be chosen, because we never indicated that the options are only inside Environment shells. Note that the jovyan prompt was never shown.

The change that fixes this case is that we auto select the "connect" option for the case when multiple environment shells are active. That feels more consistent with the auto connect action we take when there's only a single environment shell.

> I think the current indication of the Environment shell is confusing (especially as a new user who doesn't know what Environment shells). How about: (Environment: Python-2-typer)?

After discussion decided not to change the prompt.

Added the warning message for option 4. Probably better to overcommunicate.
